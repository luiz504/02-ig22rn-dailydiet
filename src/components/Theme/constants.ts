import { theme } from '~/styles'

export type ThemeVariant = 'white' | 'green' | 'gray' | 'red'

export const themeVariants: Record<ThemeVariant, string> = {
  gray: theme.colors['gray-300'],
  white: theme.colors['gray-100'],
  green: theme.colors['green-100'],
  red: theme.colors['red-100'],
} as const
