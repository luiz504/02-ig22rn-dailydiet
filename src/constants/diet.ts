export const IN_DIET_THRESHOLD = 70
