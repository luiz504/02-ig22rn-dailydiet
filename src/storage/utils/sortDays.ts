export const sortDays = (array: string[]) => {
  return array.sort((a, b) => {
    return new Date(b).getTime() - new Date(a).getTime()
  })
}
