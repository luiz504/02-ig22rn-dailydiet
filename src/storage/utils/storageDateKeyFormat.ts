import { format } from 'date-fns'

export function storageDateKeyFormat(date: Date): string {
  return format(date, 'yyyy/MM/dd')
}
