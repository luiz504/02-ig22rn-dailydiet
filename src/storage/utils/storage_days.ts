import AsyncStorage from '@react-native-async-storage/async-storage'
import { DAY_COLLECTION } from '../commons'

export async function getStoredDays() {
  return AsyncStorage.getItem(DAY_COLLECTION).then((storedData) =>
    storedData ? (JSON.parse(storedData) as string[]) : null,
  )
}

export async function setStoredDays(dayCollection: string[]) {
  return AsyncStorage.setItem(DAY_COLLECTION, JSON.stringify(dayCollection))
}
