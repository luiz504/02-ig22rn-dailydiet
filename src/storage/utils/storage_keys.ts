import { format, parse } from 'date-fns'

const FORMAT = 'yyyy/MM/dd'
export function storageDateKeyFormat(date: Date): string {
  return format(date, FORMAT)
}

export function storageKeyToDateParse(key: string) {
  return parse(key, FORMAT, new Date())
}
