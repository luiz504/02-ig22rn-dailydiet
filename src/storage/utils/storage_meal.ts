import AsyncStorage from '@react-native-async-storage/async-storage'
import { MEAL_COLLECTION } from '../commons'
import { Meal } from '~/models/Meal'

export type MealGroupByDay = {
  day: string
  meals: Meal[]
}

export async function getStoredMealsByDay(key: string) {
  return AsyncStorage.getItem(`${MEAL_COLLECTION}-${key}`).then((storedData) =>
    storedData ? (JSON.parse(storedData) as MealGroupByDay) : null,
  )
}

export async function setStoredMeals(key: string, meals: Meal[]) {
  const sortedMealsByHour = meals.sort(
    (a, b) => new Date(b.date).getTime() - new Date(a.date).getTime(),
  )
  const mealsCollection: MealGroupByDay = { day: key, meals: sortedMealsByHour }

  return AsyncStorage.setItem(
    `${MEAL_COLLECTION}-${key}`,
    JSON.stringify(mealsCollection),
  )
}

export async function getStoredMealsByDays(keys: string[]) {
  const keysCombined = keys.map((k) => `${MEAL_COLLECTION}-${k}`)
  const entries = await AsyncStorage.multiGet(keysCombined)
  return entries
    .map(([, value]) => (value ? JSON.parse(value) : null))
    .filter((value) => value) as MealGroupByDay[]
}

export async function deleteStoredMealCollection(key: string) {
  return AsyncStorage.removeItem(`${MEAL_COLLECTION}-${key}`)
}
