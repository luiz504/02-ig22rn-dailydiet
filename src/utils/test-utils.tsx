import { FC, ReactNode } from 'react'
import { ThemeProvider } from 'styled-components/native'
import { render } from '@testing-library/react-native'

import { theme } from '~/styles'
import { NavigationContainer } from '@react-navigation/native'

const AllTheProviders: FC<{ children: ReactNode }> = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <NavigationContainer>{children}</NavigationContainer>
    </ThemeProvider>
  )
}

const ThemedWrapper: FC<{ children: ReactNode }> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender: typeof render = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

const renderThemed: typeof render = (ui, options) =>
  render(ui, { wrapper: ThemedWrapper, ...options })

const splitBorderColor = (borderColor: string) => ({
  borderTopColor: borderColor,
  borderRightColor: borderColor,
  borderBottomColor: borderColor,
  borderLeftColor: borderColor,
})

export * from '@testing-library/react-native'

export { customRender as render, renderThemed, splitBorderColor }
