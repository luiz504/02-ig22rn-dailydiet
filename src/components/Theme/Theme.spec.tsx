import { Text } from 'react-native'
import { StatusBar } from 'expo-status-bar'

import { render, screen } from '~/utils/test-utils'

import { ThemeVariant, themeVariants } from './constants'
import { Theme } from '.'

describe('Theme Component', () => {
  const themeVariantsKeys = Object.keys(themeVariants)

  it('should render correctly with the default variant', () => {
    render(
      <Theme>
        <Text>anything</Text>
      </Theme>,
    )

    const containerElement = screen.getByTestId('theme-container')
    expect(containerElement).toBeOnTheScreen()
    expect(containerElement.props.style).toMatchSnapshot(
      `container-style-default`,
    )

    const statusBarElement = screen.UNSAFE_getByType(StatusBar)
    expect(statusBarElement).toBeOnTheScreen()
    expect(statusBarElement.props).toMatchSnapshot()
  })

  themeVariantsKeys.forEach((variant) => {
    it(`should render correctly with the a variant assigned`, () => {
      render(
        <Theme variant={variant as ThemeVariant}>
          <Text>anything</Text>
        </Theme>,
      )

      const containerElement = screen.getByTestId('theme-container')
      expect(containerElement.props.style).toMatchSnapshot(
        `container-style-variant=${variant}`,
      )
    })
  })
})
