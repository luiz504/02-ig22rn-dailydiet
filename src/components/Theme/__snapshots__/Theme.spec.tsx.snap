// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme Component should render correctly with the a variant assigned: container-style-variant=gray 1`] = `
{
  "backgroundColor": "#DDDEDF",
  "flex": 1,
}
`;

exports[`Theme Component should render correctly with the a variant assigned: container-style-variant=green 1`] = `
{
  "backgroundColor": "#E5F0DB",
  "flex": 1,
}
`;

exports[`Theme Component should render correctly with the a variant assigned: container-style-variant=red 1`] = `
{
  "backgroundColor": "#F4E6E7",
  "flex": 1,
}
`;

exports[`Theme Component should render correctly with the a variant assigned: container-style-variant=white 1`] = `
{
  "backgroundColor": "#FAFAFA",
  "flex": 1,
}
`;

exports[`Theme Component should render correctly with the default variant 1`] = `
{
  "style": "dark",
  "translucent": true,
}
`;

exports[`Theme Component should render correctly with the default variant: container-style-default 1`] = `
{
  "backgroundColor": "#FAFAFA",
  "flex": 1,
}
`;
